### PR TITLE
[JENKINS-28335] - Git username and password binding (GSoC 2021)

### PR DESCRIPTION
**Pull request by arpoch**
## [JENKINS-28335](https://issues.jenkins.io/browse/JENKINS-28335) - Add Git Credentials binding for Username and Password

The gitUsernamePassword implementation uses the Jenkins username and password values retrieved through Credential API, to access a remote repository over HTTP protocol. The binding uses the GIT_ASKPASS environment variable to provide credentials requested by a git operation in a pipeline job/freestyle project. Executable script(sh,bat,powershell) is attached/binded to GIT_ASKPASS variable, which is invoked when asked for user’s credentials by HTTP server.

Please refer to the Project page for more details-
https://www.jenkins.io/projects/gsoc/2021/projects/git-credentials-binding/

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] Documentation in README has been updated as necessary
- [x] Online help has been added and reviewed for any new or modified fields
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by MarkEWaite** [comment]
My lightweight first pass interactive tests all passed.  Things that I checked:

* Freestyle job with a Git username / password credential binding on Linux is able to `git fetch --all` from a private repository using a shell step.  Same shell step fails if the credentials binding is not used
* Pipeline job with a Git username / password credential binding on Linux is able to `git fetch --all` from a private repository using a shell step.  Same shell step fails if the withCredentials binding is not used

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.Set;
+
+
+public class GitUsernamePasswordBinding extends MultiBinding<StandardUsernamePasswordCredentials> implements GitCredentialBindings {
+    final static private String GIT_USERNAME_KEY = "GIT_USERNAME";
+    final static private String GIT_PASSWORD_KEY = "GIT_PASSWORD";
+    final private String gitToolName;
+    final private Map<String, String> credMap = new LinkedHashMap<>();
+    private String gitToolExe = null;
+    private boolean unixNodeType;
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:37]: I think that this field should be declared `transient` so that it is not stored in the job definition.  I believe the storage in the field is used to pass a value into setRunEnvironmentVariables but should not be considered persistent across jobs.  It is computed based on the launcher used for the current job and the computed value is used immediately to prepare the environment for the credential binding.
```suggestion
    private transient boolean unixNodeType;
```

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    @Override
+    public void setKeyBindings(@NonNull StandardCredentials credentials) {
+        credMap.put(GIT_USERNAME_KEY, ((StandardUsernamePasswordCredentials) credentials).getUsername());
+        credMap.put(GIT_PASSWORD_KEY, ((StandardUsernamePasswordCredentials) credentials).getPassword().getPlainText());
+    }
+
+    @Override
+    public void setRunEnvironmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        if (unixNodeType && ((CliGitAPIImpl) getGitClientInstance(gitToolExe,null,
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:100]: Doesn't this risk a class cast exception if the getGitClientInstance returns a JGitAPIImpl instance?

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        }
+
+        @Override
+        protected FilePath write(StandardUsernamePasswordCredentials credentials, FilePath workspace)
+                throws IOException, InterruptedException {
+            FilePath gitEcho;
+              //Hard Coded platform dependent newLine
+            if (this.unixNodeType) {
+                gitEcho = workspace.createTempFile("auth", ".sh");
+                // [#!/usr/bin/evn sh] to be used if required, could have some corner cases
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:138]:
```suggestion
                // [#!/usr/bin/env sh] to be used if required, could have some corner cases
```

File: src/main/resources/jenkins/plugins/git/Messages.properties [hunk 4]
```diff
@@ -25,3 +25,4 @@ GitSCMSource.DisplayName=Git
 GitStep.git=Git
 within.Repository=Within Repository
 additional=Additional
+GitUsernamePasswordBind.DisplayName=Git Username and Password
```
Comment (MarkEWaite) [inline src/main/resources/jenkins/plugins/git/Messages.properties:28]: I don't see any mention of `GitUsernamePasswordBind` anywhere else in the code.  Shouldn't this be removed or replaced with a reference to the current class name?
```suggestion
GitUsernamePasswordBind.DisplayName=Git Username and Password
```

**Review by MarkEWaite — COMMENTED** [review]
Initial comments and questions

File: src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<div>
+    Set the git username / password credential for <strong>HTTP</strong> and <strong>HTTPS</strong> protocols.
+
+<p>
+<strong>Shell example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Comment (MarkEWaite) [inline src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:7]: Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

File: src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  sh 'git fetch --all'
+}
+</pre>
+</p>
+
+<p>
+<strong>Batch example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Comment (MarkEWaite) [inline src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:16]: Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

File: src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  bat 'git submodule update --init --recursive'
+}
+</pre>
+</p>
+
+<p>
+<strong>Powershell example</strong>
+<pre>
+withCredentials([GitUsernamePassword('my-credentials-id')]) {
```
Comment (MarkEWaite) [inline src/main/resources/jenkins/plugins/git/GitUsernamePasswordBinding/help-credentialsId.html:25]: Needs the `credentialsId` argument explicitly named
```suggestion
withCredentials([GitUsernamePassword(credentialsId: 'my-credentials-id')]) {
```

**Comment by arpoch** [comment]
> Enabling JGit and dragging it to be first in the list of implementations will cause the build log to report that JGit is not supported with the credential binding, even if the git tool name is 'Default' or 'git-2.32.0' (a tool that I defined). We should allow JGit to be selected as the first git implementation as is done on ci.jenkins.io.

The issues is due to this [line](https://github.com/jenkinsci/git-plugin/blob/b4fdcf9a4829ef72a27b37af451a187befabd221/src/main/java/jenkins/plugins/git/GitCredentialBindings.java#L38),  which checks if the installation is `GitTool` type specific and not `JGitTool` or `JGitApache` type specific, the assumption made here of selecting a tool at index 0 is what lead to this situation. 
I have made some fixes and things are back on track.

Since your binding will only work for Cli Git specific git tool so its better to convert the method [getGitTool](https://github.com/jenkinsci/git-plugin/blob/b4fdcf9a4829ef72a27b37af451a187befabd221/src/main/java/jenkins/plugins/git/GitCredentialBindings.java#L28) to `getCliGitTool` returning a git tool which is both cli git type specific and name specific provided by the user. So returning a generic git tool as done by `resloveGitTool` won't sever our entire purpose.

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java (hunk 2 above)
Comment (arpoch) [reply src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:100]: As mentioned in this [comment](https://github.com/jenkinsci/git-plugin/pull/1104#issuecomment-877800361) the method `getCliGitTool` is the key to prevent this risk.
Also the reason to create getCliGitTool API was to ensure that the binding will only support CLI Git and no other git impls.

**Comment by MarkEWaite** [comment]
@arpoch could you merge the master branch into this branch so that the incremental builds generated by ci.jenkins.io will be stored in the Jenkins artifact repository?  That will allow me to insert the incremental version number into my plugins.txt file and use it to generate my Jenkins configuration from code.

**Comment by MarkEWaite** [comment]
All my functionality checks pass with the latest code changes (as of b1336813ef9ab9f3f827517cc9cb67743eff738b) .  Well done!

I see the pick list of git tools as well and am able to select one.  Very nice.

Unfortunately, when I see the pick list, it also always shows the message that the default git tool will be used, no matter which git tool I select.  Is that a mistaken check of the field contents or something else?

screencapture-mark-pc2-markwaite-net-8080-job-Hosting-Providers-job-github-job-github-bin-https-withCredentials-windows-pipeline-syntax-2021-07-16-17_44_35

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        public FormValidation doCheckGitToolName(@QueryParameter String toolName){
+            if(toolName == null){
+                return FormValidation.warning("Default Git Tool(Git/JGit) will be used");
+            }else {
+                return FormValidation.ok();
+            }
+        }
+
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:194]: I believe this code can be removed because there is no need to check for valid content when showing a pick list.  The pick list can only contain valid content.
```suggestion
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]

**Review by MarkEWaite — APPROVED** [review]
Needs documentation in the README, but otherwise looks ready to merge and release.

The README documentation can be in a separate pull request if needed.

@rishabhBudhouliya and @justinharringa , any concerns from you that should be addressed before this is merged and released?

File: README.adoc [hunk 9]
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
 
 image:images/multibranch-pipeline.png[link=https://youtu.be/B_2FXWI6CWg]
 
+[#credential-binding]
+=== Git Credentials Binding
+
+The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *SSH* protocols using command line git in a Pipeline job.
```
Comment (MarkEWaite) [inline README.adoc:57]: SSH support is not yet ready.  Let's not mention it yet.
```suggestion
The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *HTTPS* protocols using command line git in a Pipeline job.
```

File: README.adoc [hunk 10]
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
...
+
+The git credential bindings are accessible through the `withCredentials` step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.
```
Comment (MarkEWaite) [inline README.adoc:59]: Hyperlink to the step documentation?
```suggestion
The git credential bindings are accessible through the link:https://www.jenkins.io/doc/pipeline/steps/credentials-binding/#withcredentials-bind-credentials-to-variables[`withCredentials`] step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.
```

File: README.adoc [hunk 11]
```diff
@@ -51,6 +51,48 @@ The link:https://youtu.be/B_2FXWI6CWg[30 minute video clip] below introduces Mul
...
+This binding provides authentication support over *HTTP* protocol using command line git in a Pipeline job.
+
+Procedure::
+
+. Click the Pipeline Syntax _Snippet Generator_ and choose the `withCredentials` step, add Git Username and Password binding.
+. Choose the required credentials and Git tool name, specific to the generated Pipeline snippet.
+
+image:images/git-credentials-usernamepassword-binding-pipline-job.png[Git-Username-and Password-Binding-Pipeline-Job]
+
+NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.
```
Comment (MarkEWaite) [inline README.adoc:73]: This seems like an interesting implementation detail, but placing it in a NOTE gives it more emphasis that I think is justified.  The user should not need to do anything with either of these variables.  They are internal implementation details.
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to `sh`, `bat`, and `powershell` steps inside the `withCredentials` block of a Pipeline job.
```

File: README.adoc [hunk 12]
```diff
@@ -544,6 +586,25 @@ Project Name in ViewGit::
...
 
+[#git-bindings]
+== Git Bindings
+
+The git plugin provides one binding to support authenticated git opreations over *HTTP* or * HTTPS* protocol, namely `Git Username and Password`.
+The git plugin depeneds on the Credential Binding Plugin to support these bindings.
+
+To access the `Git Username and Password` binding in a pipeline job, visit <<credential-binding>>
+
+Incase of a FreeStyle project the steps are as follows:
```
Comment (MarkEWaite) [inline README.adoc:597]: Remove the 'Incase' word
```suggestion
Freestyle projects can use git credential binding with the following steps:
```

File: README.adoc [hunk 13]
```diff
@@ -544,6 +586,25 @@ Project Name in ViewGit::
...
+
+. Check the box _Use secret text(s) or file(s)_, add Git Username and Password binding.
+
+. Choose the required credentials and Git tool name.
+
+image:images/git-credentials-usernamepassword-binding-freestyle-project.png[Git-Username-and Password-Binding-Freestyle-project]
+
+NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.
```
Comment (MarkEWaite) [inline README.adoc:605]: Remove the NOTE: emphasis here as well
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to shell, batch, and powershell steps in a Freestyle job.
```

**Review by MarkEWaite — APPROVED** [review]
Minor phrasing changes proposed for your consideration.  Optional

File: src/main/java/jenkins/plugins/git/GitCredentialBindings.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+     **/
+    default GitTool getCliGitTool(Run<?, ?> run, String gitToolName,
+                              TaskListener listener) throws IOException, InterruptedException {
+
+        Executor buildExecutor = run.getExecutor();
+        if (buildExecutor != null) {
+            Node currentNode = buildExecutor.getOwner().getNode();
+            //Check node is not null
+            if (currentNode != null) {
+                GitTool nameSpecificGitTool = GitUtils.resolveGitTool(gitToolName,currentNode,new EnvVars(),listener);
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitCredentialBindings.java:72]: nitpick: please follow conventional formatting rules (whitespace after comma)

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+            GenerateGitScript gitScript = new GenerateGitScript(
+                                                  credentials.getUsername(),
+                                                  credentials.getPassword().getPlainText(),
+                                                  credentials.getId(),
+                                                  this.unixNodeType);
+            FilePath gitTempFile = gitScript.write(credentials, unbindTempDir.getDirPath());
+            secretValues.put("GIT_ASKPASS", gitTempFile.getRemote());
+            return new MultiEnvironment(secretValues, publicValues, unbindTempDir.getUnbinder());
+        } else {
+            taskListener.getLogger().println("JGit and JGitApache type Git tools are not supported by this binding");
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:87]: `getCliGitTool` could return a null value for a lot of other reasons than just the possibility of git tool being JGit and JGitApache types. I suggest you should make this log statement more general: 

"Unable to perform binding because git tool resolution has failed"

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        keys.add(GIT_PASSWORD_KEY);
+        return keys;
+    }
+
+    @Override
+    public void setCredentialPairBindings(@NonNull StandardCredentials credentials,Map<String,String> publicValues, Map<String,String> secretValues) {
+        StandardUsernamePasswordCredentials usernamePasswordCredentials = (StandardUsernamePasswordCredentials) credentials;
+        if(usernamePasswordCredentials.isUsernameSecret()){
+            secretValues.put(GIT_USERNAME_KEY, usernamePasswordCredentials.getUsername());
+        }else{
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:105]: nitpick: formatting (whitespace)

File: src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+            publicValues.put(GIT_USERNAME_KEY, usernamePasswordCredentials.getUsername());
+        }
+        secretValues.put(GIT_PASSWORD_KEY, usernamePasswordCredentials.getPassword().getPlainText());
+    }
+
+    /*package*/void setGitEnvironmentVariables(@NonNull GitClient git, Map<String,String> publicValues) throws IOException, InterruptedException {
+        setGitEnvironmentVariables(git,null,publicValues);
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java:112]: nitpick: formatting

File: src/main/java/jenkins/plugins/git/GitCredentialBindings.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Use GitClient to perform git operations in a repository. Using Git implementations JGit/JGit Apache/Cli Git
+     * @param gitExe The path {@link java.lang.String} to git executable {@link org.jenkinsci.plugins.gitclient.Git#using(String)}
+     * @param repository The path {@link java.lang.String} to working directory {@link org.jenkinsci.plugins.gitclient.Git#in(File)}
+     * @param env The environment values {@link hudson.EnvVars}
+     * @param listener The task listener.
+     * @return a GitClient implementation {@link org.jenkinsci.plugins.gitclient.GitClient}
+     **/
+    GitClient getGitClientInstance(String gitExe, FilePath repository,
```
Comment (rishabhBudhouliya) [inline src/main/java/jenkins/plugins/git/GitCredentialBindings.java:46]: why is this added in the contract to be implemented? Can we not provide the logic assuming the implementation does not change for any reason?

Comment (rishabhBudhouliya) [reply src/main/java/jenkins/plugins/git/GitCredentialBindings.java:46]: similar to getCliGitTool

**Review by rishabhBudhouliya — APPROVED** [review]
I've left some fairly minor comments, nice work.

File: src/main/java/jenkins/plugins/git/GitCredentialBindings.java (hunk 18 above)
Comment (jglick) [reply src/main/java/jenkins/plugins/git/GitCredentialBindings.java:46]: (Still unexplained five years later.)
